File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/7b63345be437e04c0f3b18ccf1e0ee68.py
import sys
from itertools import accumulate

class DSum:
    class Solver:
        def __init__(self, in_stream, out_stream):
            self._in = in_stream
            self._out = out_stream

        def rd(self):
            return map(int, self._in.readline().split())

        def pr(self, *args):
            self._out.write(' '.join(map(str, args)) + '\n')

        def add_item(self, knapsack, size, sum_):
            for k in range(K - size, -1, -1):
                knapsack[k + size] = max(knapsack[k + size], knapsack[k] + sum_)

        def recurse(self, start, end, knapsack):
            if start >= end:
                return

            if end - start == 1:
                sum_ = 0
                self.best = max(self.best, knapsack[K])

                for prefix in range(1, len(A[start]) + 1):
                    sum_ += A[start][prefix - 1]
                    self.best = max(self.best, sum_ + knapsack[K - prefix])

                return

            mid = (start + end) // 2
            state = knapsack[:]

            for i in range(start, mid):
                self.add_item(state, len(A[i]), self.A_sum[i])

            self.recurse(mid, end, state)
            state = knapsack[:]

            for i in range(mid, end):
                self.add_item(state, len(A[i]), self.A_sum[i])

            self.recurse(start, mid, state)

        def solve(self):
            global N, K, A, A_sum
            N, K = self.rd()
            A = []
            A_sum = []

            for _ in range(N):
                t, *a = self.rd()
                A.append(a)
                if t > K:
                    A[-1] = A[-1][:K]
                A_sum.append(sum(A[-1]))

            self.best = 0
            knapsack = [0] * (K + 1)
            self.recurse(0, N, knapsack)
            self.pr(self.best)

    @staticmethod
    def solve(in_stream, out_stream):
        DSum.Solver(in_stream, out_stream).solve()

if __name__ == "__main__":
    input_stream = sys.stdin
    output_stream = sys.stdout
    DSum.solve(input_stream, output_stream)

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/7b63345be437e04c0f3b18ccf1e0ee68.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>

#define MAX_N 100000
#define MAX_K 100

int64_t A[MAX_N][MAX_K];
int64_t A_sum[MAX_N];
int64_t best;

void add_item(int64_t knapsack[MAX_K], int size, int64_t sum) {
    for (int k = MAX_K - size; k >= 0; k--) {
        knapsack[k + size] = knapsack[k + size] > knapsack[k] + sum ? knapsack[k + size] : knapsack[k] + sum;
    }
}

void recurse(int start, int end, int64_t knapsack[MAX_K]) {
    if (start >= end) {
        return;
    }

    if (end - start == 1) {
        int64_t sum = 0;
        best = best > knapsack[MAX_K - 1] ? best : knapsack[MAX_K - 1];

        for (int prefix = 1; prefix <= MAX_K; prefix++) {
            sum += A[start][prefix - 1];
            best = best > sum + knapsack[MAX_K - prefix] ? best : sum + knapsack[MAX_K - prefix];
        }

        return;
    }

    int mid = (start + end) / 2;
    int64_t state[MAX_K];

    for (int i = start; i < mid; i++) {
        add_item(state, MAX_K, A_sum[i]);
    }

    recurse(mid, end, state);
    for (int i = mid; i < end; i++) {
        add_item(state, MAX_K, A_sum[i]);
    }

    recurse(start, mid, state);
}

void solve() {
    int N, K;
    scanf("%d %d", &N, &K);

    for (int i = 0; i < N; i++) {
        int t;
        scanf("%d", &t);

        for (int j = 0; j < t; j++) {
            scanf("%lld", &A[i][j]);
        }

        if (t > K) {
            t = K;
        }

        A_sum[i] = 0;
        for (int j = 0; j < t; j++) {
            A_sum[i] += A[i][j];
        }
    }

    best = 0;
    int64_t knapsack[MAX_K] = {0};
    recurse(0, N, knapsack);
    printf("%lld\n", best);
}

int main() {
    solve();
    return 0;
}
